### PR TITLE
Add call duration to call view.

### PR DESCRIPTION
**Pull request by charlesmchen**
Show call duration for connected calls.

PTAL @michaelkirk 

screen shot 2017-01-13 at 5 30 21 pm

### CODE REVIEW

File: Signal/src/view controllers/CallViewController.swift [hunk 1]
```diff
@@ -506,6 +528,24 @@ class CallViewController: UIViewController, CallDelegate {
 
         default: break
         }
+
+        if callState == .connected {
+            if callDurationTimer == nil {
+                let kDurationUpdateFrequencySeconds = 1 / 20.0
+                callDurationTimer = Timer.scheduledTimer(timeInterval: kDurationUpdateFrequencySeconds,
```
Comment (charlesmchen) [inline Signal/src/view controllers/CallViewController.swift]: Wait, this is a retain cycle. Working on it...

Comment (charlesmchen) [inline Signal/src/view controllers/CallViewController.swift]: Fixed it.

File: Signal/src/view controllers/CallViewController.swift [hunk 2]
```diff
@@ -472,7 +473,28 @@ class CallViewController: UIViewController, CallDelegate {
         case .answering:
             return NSLocalizedString("IN_CALL_SECURING", comment: "Call setup status label")
         case .connected:
-            return NSLocalizedString("IN_CALL_TALKING", comment: "Call setup status label")
+            if let call = self.call {
+                let callDuration = call.connectionDuration()
+                let callDurationDate = Date(timeIntervalSinceReferenceDate:callDuration)
+                let dateFormatter = DateFormatter()
```
Comment (pouriaalmassi) [inline Signal/src/view controllers/CallViewController.swift]: Would it be worth caching this DateFormatter? https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/DataFormatting/Articles/dfDateFormatting10_4.html#//apple_ref/doc/uid/TP40002369-SW10

File: Signal/src/view controllers/CallViewController.swift [hunk 3]
```diff
@@ -506,6 +528,37 @@ class CallViewController: UIViewController, CallDelegate {
 
         default: break
         }
+
+        if callState == .connected {
+            if callDurationTimer == nil {
+                let kDurationUpdateFrequencySeconds = 1 / 20.0
```
Comment (michaelkirk) [inline Signal/src/view controllers/CallViewController.swift]: Just curious, how'd you arrive at 20x per second?

File: Signal/src/view controllers/CallViewController.swift [hunk 4]
```diff
@@ -472,7 +473,28 @@ class CallViewController: UIViewController, CallDelegate {
         case .answering:
             return NSLocalizedString("IN_CALL_SECURING", comment: "Call setup status label")
         case .connected:
-            return NSLocalizedString("IN_CALL_TALKING", comment: "Call setup status label")
+            if let call = self.call {
+                let callDuration = call.connectionDuration()
+                let callDurationDate = Date(timeIntervalSinceReferenceDate:callDuration)
```
Comment (michaelkirk) [inline Signal/src/view controllers/CallViewController.swift]: style nit: space after `:`

File: Signal/src/call/SignalCall.swift [hunk 5]
```diff
@@ -41,6 +41,16 @@ protocol CallDelegate: class {
     var state: CallState {
         didSet {
             Logger.debug("\(TAG) state changed: \(oldValue) -> \(state)")
+
+            // Update connectedDate
+            if state == .connected {
+                if connectedDate == nil {
+                    connectedDate = NSDate()
+                }
+            } else {
```
Comment (michaelkirk) [inline Signal/src/call/SignalCall.swift]: Will this immediately wipe the call duration when the call ends?

ux nit: The call view hangs out for a second when the call was hung up by the remote party. I think we should continue to display the time (frozen) after the other party hangs up.

File: Signal/src/view controllers/CallViewController.swift [hunk 6]
```diff
@@ -506,6 +528,37 @@ class CallViewController: UIViewController, CallDelegate {
...
+            callDurationTimer?.invalidate()
+            callDurationTimer = nil
+        }
+    }
+
+    func updateCallDuration(timer: Timer?) {
+        updateCallUI(callState: call.state)
+    }
+
+    override func viewDidDisappear(_ animated: Bool) {
```
Comment (michaelkirk) [inline Signal/src/view controllers/CallViewController.swift]: style nit: I like to keep the view lifecyle methods together at the top, after any inits.

File: Signal/src/view controllers/CallViewController.swift (hunk 2 above)
Comment (michaelkirk) [inline Signal/src/view controllers/CallViewController.swift]: Consider moving all the formatting to the model and out of the VC. Keeps the VC lighter, and makes it easily testable.

Comment (pouriaalmassi) [inline Signal/src/view controllers/CallViewController.swift]: Forgive me if I'm going against norms established by this project but keeping formatting at a higher level might easy access to formatters for other models and VCs. Consider the example laid out [here](http://sandmoose.com/post/119962774817/caching-nsdateformatter) and a cautionary explanation on minimizing creation of formatters by bbum [here](http://stackoverflow.com/a/4442389). 

Creating a formatter cache would be out of scope for this PR so a separate PR may be advised.

Comment (charlesmchen) [inline Signal/src/view controllers/CallViewController.swift]: I've moved this `DateFormatter` to a lazily-populated class property. Formatting feels like a view concern, so I'd rather it didn't leak into the model.  

Thanks @pouriaalmassi @michaelkirk

File: Signal/src/view controllers/CallViewController.swift (hunk 4 above)
Comment (charlesmchen) [inline Signal/src/view controllers/CallViewController.swift]: Done. Too bad clang-format doesn't catch this. I can't find a corresponding style option. http://clang.llvm.org/docs/ClangFormatStyleOptions.html

**Comment by charlesmchen** [comment]
> Consider moving all the formatting to the model and out of the VC. Keeps the VC lighter, and makes it easily testable.

Formatting seems like a view concern, and I'd rather not leak it into the model. Some people use ViewModels for this kind of logic, but this is just a handful of lines, so it doesn't seem worthwhile. I'm personally much more concerned with keeping models "dumb" and "light" than VCs.

File: Signal/src/view controllers/CallViewController.swift (hunk 3 above)
Comment (charlesmchen) [inline Signal/src/view controllers/CallViewController.swift]: It was a bit arbitrary.  I just wanted the frame rate to be fast enough that the ticking of seconds would feel smooth and regular.

File: Signal/src/view controllers/CallViewController.swift (hunk 6 above)
Comment (charlesmchen) [inline Signal/src/view controllers/CallViewController.swift]: +1 Done.

File: Signal/src/call/SignalCall.swift (hunk 5 above)
Comment (charlesmchen) [inline Signal/src/call/SignalCall.swift]: >I think we should continue to display the time (frozen) after the other party hangs up.

I'm using the same UILabel to display the "call status" (Dialing, Call Ended) and the "call duration", just like the iOS call screen.   The "call duration" is shown while the call is connected. When the call ends, the call status changes to something like "Call Ended," again, just like the iOS call screen.  I like how the iOS call screen works - what's the argument for deviating?

File: Signal/src/view controllers/CallViewController.swift (hunk 4 above)
Comment (michaelkirk) [inline Signal/src/view controllers/CallViewController.swift]: I'm not sure that clang format supports swift. I believe the swift linter catches this though.

**Pull request merged**